src/lndk_offers.rs

```diff
@@ -1,7 +1,237 @@
+use async_trait::async_trait;
+use bitcoin::hashes::sha256::Hash;
+use bitcoin::network::constants::Network;
+use bitcoin::secp256k1::schnorr::Signature;
+use bitcoin::secp256k1::{Error as Secp256k1Error, PublicKey};
+use futures::executor::block_on;
+use lightning::offers::invoice_request::{InvoiceRequest, UnsignedInvoiceRequest};
+use lightning::offers::merkle::SignError;
 use lightning::offers::offer::Offer;
-use lightning::offers::parse::Bolt12ParseError;
+use lightning::offers::parse::{Bolt12ParseError, Bolt12SemanticError};
+use std::error::Error;
+use std::fmt::Display;
+use tokio::task;
+use tonic_lnd::signrpc::{KeyLocator, SignMessageReq};
+use tonic_lnd::tonic::Status;
+use tonic_lnd::Client;
+
+#[derive(Debug)]
+/// OfferError is an error that occurs during the process of paying an offer.
+pub(crate) enum OfferError<Secp256k1Error> {
+    BuildUIRFailure(Bolt12SemanticError),
+    SignFailure(SignError<Secp256k1Error>),
+}
+
+impl Display for OfferError<Secp256k1Error> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OfferError::BuildUIRFailure(e) => write!(f, "{:?}", e),
+            OfferError::SignFailure(e) => write!(f, "{:?}", e),
+        }
+    }
+}
+
+impl Error for OfferError<Secp256k1Error> {}
 
 // Decodes a bech32 string into an LDK offer.
 pub fn decode(offer_str: String) -> Result<Offer, Bolt12ParseError> {
     offer_str.parse::<Offer>()
 }
+
+#[allow(dead_code)]
+// create_request_invoice builds and signs an invoice request, the first step in the BOLT 12 process of paying an offer.
+pub(crate) async fn create_request_invoice(
+    mut signer: impl MessageSigner + std::marker::Send + 'static,
+    offer: Offer,
+    metadata: Vec<u8>,
+    network: Network,
+    msats: u64,
+) -> Result<InvoiceRequest, OfferError<bitcoin::secp256k1::Error>> {
+    let key_loc = KeyLocator {
+        key_family: 6,
+        key_index: 1,
+    };
+
+    let pubkey_bytes = signer
+        .derive_key(key_loc.clone())
+        .await
+        .expect("failed to get key");
```

Comment:
Let's use `Result` rather than `expect` - this will panic if LND is down, which I think is a reasonable error case we want to handle.